### PR TITLE
#334: Add prefix_len property to LocalFolder.

### PR DESCRIPTION
**Pull request by marchevskiy**
This PR attempts to resolve an edge case in the prefix length derivation of the LocalFolder class by using `os.path.splitdrive` to generalize cases when the root path is essentially both a leading and trailing delimiter. 

The unit tests are passing for me but I have not been able to test the windows compatibility. If the CI checks pass I will work on getting a VM setup.

### CODE REVIEW

File: b2/sync/folder.py [hunk 1]
```diff
@@ -70,14 +70,52 @@ def __init__(self, root):
...
+        if len(tail) == 1:
+            return len(drive) + len(tail)
+        else:
+            return len(self.root) + 1
 
     def folder_type(self):
         return 'local'
 
     def all_files(self, reporter):
-        prefix_len = len(self.root) + 1  # include trailing '/' in prefix length
```
Comment (bwbeach) [inline b2/sync/folder.py]: We've lost the comment saying that the trailing "/" is in the prefix length.  I think that would be worth saying somewhere.

It's probably also worth saying that self.root has a trailing "/" only if it's the root of the filesystem.

File: test/test_sync.py [hunk 2]
```diff
@@ -105,6 +107,58 @@ def test_invalid_permissions(self):
...
+            # Build a folder.
+            folder = LocalFolder(base_path)
+
+            # Verify that the prefix truncates the root path correctly.
+            self.assertEqual(folder.root[folder.prefix_len:], u'')
+
+            # Verify full path to relative path conversion.
+            test_path = folder.root
+            test_sub_path = u'test'
+            if not folder.root.endswith(separator):
```
Comment (bwbeach) [inline test/test_sync.py]: This "if" is what os.path.join does:

```
>>> os.path.join('/foo/', 'bar')
'/foo/bar'
>>> os.path.join('/foo', 'bar')
'/foo/bar'
```

File: test/test_sync.py [hunk 3]
```diff
@@ -105,6 +107,58 @@ def test_invalid_permissions(self):
                 os.path.join(tmpdir, self.NAMES[0])
             )
 
+    def test_prefix_calculation(self):
+        """
```
Comment (bwbeach) [inline test/test_sync.py]: I like the comments.  Thorough and easy to read.

File: test/test_sync.py [hunk 4]
```diff
@@ -105,6 +107,58 @@ def test_invalid_permissions(self):
...
+
+            For windows, case 1) is different (e.g. C:\\) but the others should be the same.
+
+        """
+
+        prefix = os.path.normpath(u'C://') if platform.system() == 'Windows' else unicode(os.path.sep)
+        separator = unicode(os.path.sep)
+        root = os.path.normpath(os.getcwdu())  # Always has leading but not trailing separator.
+        relative = u''.join(choice(lowercase) for _ in range(randint(4, 8)))
+        permutations = dict(**{
```
Comment (bwbeach) [inline test/test_sync.py]: How about just `{ ... }` instead of `dict(**{ ... })`?

File: test/test_sync.py (hunk 2 above)
Comment (marchevskiy) [reply test/test_sync.py]: Yep good catch.

File: test/test_sync.py (hunk 4 above)
Comment (marchevskiy) [reply test/test_sync.py]: Yep sorry, got fooled by my IDE there.

File: b2/sync/folder.py (hunk 1 above)
Comment (marchevskiy) [reply b2/sync/folder.py]: I'll add it back in. There's somewhat of an explanation within the docstring for the `prefix_len` property, but it is not as explicit. 

Thinking about this a bit more, I believe there's no reason to keep passing `prefix_len` to `_walk_relative_paths` since it's a property of the class now.

File: b2/sync/folder.py [hunk 5]
```diff
@@ -70,14 +70,52 @@ def __init__(self, root):
         """
         if not isinstance(root, six.text_type):
             raise ValueError('folder path should be unicode: %s' % repr(root))
-        self.root = os.path.abspath(root)
+
+        # Using normpath here allows windows and linux paths to be handled.
+        self.root = os.path.normpath(os.path.abspath(root))
```
[the remark this answers is not in the document]
Comment (marchevskiy) [reply b2/sync/folder.py]: Yeah I was a bit confused by the documentation here, and think I got it backwards. It's definitely a no-op here and I'll remove it.

**Comment by marchevskiy** [comment]
The Travis CI checks appear to be failing because string.lowercase doesn't exist for python3. I'll find a way to use the six module or just generate a random string differently.

**Comment by svonohr** [comment]
By chance I thought about this bug already months ago and came up with a different solution:
`prefix_len = len(os.path.join(self.root, ''))`
By appending an empty string the join will add a trailing slash to the path, if one does not already exist, or if its the root of the filesystem it also won't add another slash. I'm not sure however, if this covers all cases this PR covers or if this behavior of `join` is standard.

**Comment by marchevskiy** [comment]
That's certainly a more elegant solution. I'm new to the github ecosystem so I'm not sure if you can just commit to my branch?

**Comment by svonohr** [comment]
You'd have to add me as a collaborator to your fork. Or I could fork your branch, do the changes in my branch and then create a new pull request. It's probably easier if you just change it in your branch (and I don't have much time at the moment).

**Comment by marchevskiy** [comment]
Ok just wanted to make sure you get credit. I'll update this branch and hopefully it will be ready to merge.

**Comment by svonohr** [comment]
I haven't tested your branch yet, but I believe there is another place which doesn't handle the root path properly. In sync/folder_parser.py line 30 we have the following code:
```
if folder_name.endswith('/'):
    folder_name = folder_name[:-1]
return LocalFolder(folder_name)
```
This will just remove the whole string when it's just the root path.

**Comment by marchevskiy** [comment]
Yeah it will end up using the current working directory instead of the path in `folder_name`. The fix is trivial, just let `LocalFolder` take care of the root prefix and don't pre-process in higher level functions. 

I think it might be worthwhile to write some integration unit tests that mock out os.path* to get better coverage in this area. I can take a shot at drafting something up for that in a different branch.

**Comment by marchevskiy** [comment]
I stumbled on an issue when testing this in my VM where some files had relative symlinks - e.g. `/usr/bin/X11 -> .` - and would cause an infinite recursion loop in `_walk_relative_paths` so I added a bit of protective code against that case.

**Comment by svonohr** [comment]
Isn't that a general problem with cyclic symlinks? I tried synching such a cyclic directory and b2 cli builds a path of arround 500 characters and then complaints that the symlink seems to be broken.

**Comment by marchevskiy** [comment]
Yeah the validate closure had some issues. Can you please provide an example of the cyclical directory structure you were able to produce a broken symlink with?

**Comment by svonohr** [comment]
You don't even have to use relative links. I've created some directories `/foo/bar` and then added an absolute link to `/foo`. This way b2 cli will come up with paths like `/foo/bar/link/bar/link/...`. I'm not sure how one would prevent this.

**Comment by ppolewicz** [comment]
[os.path.islink](https://docs.python.org/3/library/os.path.html#os.path.islink) and [os.path.realpath](https://docs.python.org/3/library/os.path.html#os.path.realpath), maybe [os.path.commonprefix](https://docs.python.org/3/library/os.path.html#os.path.commonprefix)

**Comment by marchevskiy** [comment]
@svonohr I believe that will not occur on this branch, perhaps you were on the master branch?

Master:
``` 
root@2db87b17b227:/# mkdir -p /foo/bar/
root@2db87b17b227:/# ln -s /foo /foo/bar/link
root@2db87b17b227:/# b2 sync --dryRun /foo/ b2://bucket-74358734
WARNING: /foo/bar/link/bar/link/bar/link/bar/link/bar/link/bar/link/bar/link/bar/link/bar/link/bar/link/bar/link/bar/link/bar/link/bar/link/bar/link/bar/link/bar/link/bar/link/bar/link/bar/link/bar/link/bar/link/bar/link/bar/link/bar/link/bar/link/bar/link/bar/link/bar/link/bar/link/bar/link/bar/link/bar/link/bar/link/bar/link/bar/link/bar/link/bar/link/bar/link/bar/link/bar/link could not be accessed (broken symlink?)
```

issues/334:
```
root@2db87b17b227:/# mkdir -p /foo/bar/
root@2db87b17b227:/# ln -s /foo /foo/bar/link
root@2db87b17b227:/# b2 sync --dryRun /foo/ b2://bucket-74358734
upload bar/link      
```

...though if I do actually perform the sync, I'll get an exception thrown for an invalid source: 

`b2_upload(/foo/bar/link, bar/link, 1507226301968): InvalidUploadSource('/foo/bar/link',) Invalid upload source: /foo/bar/link`. 

I'm not *exactly* sure how symlinks are intended to be handled, so I could use some help from the developers in understanding what the expected behavior is.

**Comment by bwbeach** [comment]
Thanks, @marchevskiy, for the all the work you did on this.  Your work made it clear that my initial approach on this code was just wrong.  I tried having the code keep two paths: the path on the local disk, and the B2 path.  That made things a lot clearer.  See #384.
